File: Memory Optimization/sol1_gpt.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int T = Integer.parseInt(br.readLine().trim());
        StringBuilder output = new StringBuilder();

        while (T-- > 0) {
            int N = Integer.parseInt(br.readLine().trim());

            StringTokenizer st = new StringTokenizer(br.readLine());

            HashMap<Integer, Integer> maxExponent = new HashMap<>();

            for (int i = 0; i < N; i++) {
                int x = Integer.parseInt(st.nextToken());

                int exponent = 0;

                while (x % 2 == 0) {
                    x /= 2;
                    exponent++;
                }

                int oldValue = maxExponent.getOrDefault(x, 0);

                if (exponent > oldValue) {
                    maxExponent.put(x, exponent);
                }
            }

            long answer = 0;

            for (int exponent : maxExponent.values()) {
                answer += exponent;
            }

            output.append(answer).append('\n');
        }

        System.out.print(output);
    }
}

File: Memory Optimization/sol1_gpt.py
import sys

input = sys.stdin.readline

T = int(input())

for _ in range(T):
    N = int(input())
    arr = list(map(int, input().split()))

    max_exponent = {}

    for x in arr:
        exponent = 0

        while x % 2 == 0:
            x //= 2
            exponent += 1

        if x not in max_exponent:
            max_exponent[x] = exponent
        else:
            max_exponent[x] = max(max_exponent[x], exponent)

    answer = sum(max_exponent.values())

    print(answer)
